### PR TITLE
code actions on save for autosave `afterDelay`

### PR DESCRIPTION
**Pull request by justschen**
ref: https://github.com/microsoft/vscode/issues/200881

This will trigger code actions for organize imports when the following conditions are applied:
- file auto save is set to `afterDelay`
- editor code actions (`source.organizeImports`) is set to `always`
- setting: `editor.codeActions.triggerOnFocusChange` is `true` (default is false)

note: will still be triggered when `afterDelay` is on and we manually trigger a safe (same as current behavior) 

TLDR: we match the behavior of `onFocusChange` or `onWindowChange` save styles when code actions are `always` and our save style is `afterDelay`

### CODE REVIEW

File: src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts [hunk 1]
```diff
@@ -273,7 +275,50 @@ class CodeActionOnSaveParticipant implements ITextFileSaveParticipant {
...
+		@IEditorService private readonly editorService: IEditorService,
+		@ICodeEditorService private readonly codeEditorService: ICodeEditorService,
+	) {
+		this.hostService.onDidChangeFocus(async () => {
+			this.triggerCodeActionsCommand();
+		});
+
+		this.editorService.onDidActiveEditorChange(async () => {
+			this.triggerCodeActionsCommand();
+		});
```
Comment (Tyriar) [inline src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts:288]: Should register these (extend `Disposable` + `_register`)

Comment (Tyriar) [inline src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts:288]: The async here doesn't do anything

File: src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts [hunk 2]
```diff
@@ -273,7 +275,50 @@ class CodeActionOnSaveParticipant implements ITextFileSaveParticipant {
...
+	}
+
+	private async triggerCodeActionsCommand() {
+		if (this.configurationService.getValue<boolean>('editor.codeActions.triggerOnFocusChange') && this.configurationService.getValue<string>('files.autoSave') === 'afterDelay') {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts:292]: Are there any constants we can use for these setting IDs instead?

File: src/vs/editor/contrib/codeAction/browser/codeActionContributions.ts [hunk 3]
```diff
@@ -45,3 +45,15 @@ Registry.as<IConfigurationRegistry>(Extensions.Configuration).registerConfigurat
 		},
 	}
 });
+
+Registry.as<IConfigurationRegistry>(Extensions.Configuration).registerConfiguration({
+	...editorConfigurationBaseNode,
+	properties: {
+		'editor.codeActions.triggerOnFocusChange': {
+			type: 'boolean',
+			scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
```
Comment (Tyriar) [inline src/vs/editor/contrib/codeAction/browser/codeActionContributions.ts:54]: Should this be `restricted`?

File: src/vs/editor/contrib/codeAction/browser/codeActionContributions.ts [hunk 4]
```diff
@@ -45,3 +45,15 @@ Registry.as<IConfigurationRegistry>(Extensions.Configuration).registerConfigurat
...
+			markdownDescription: nls.localize('triggerOnFocusChange', 'Enable triggering {0} when {1} is set to `afterDelay`. Code Actions must be set to `always` to be triggered for window and focus changes.', '`#editor.codeActionsOnSave#`', '`#files.autoSave#`'),
```
Comment (Tyriar) [inline src/vs/editor/contrib/codeAction/browser/codeActionContributions.ts:55]: `afterDelay` and `always` should be {2} and {3}

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/codeAction/browser/codeActionContributions.ts (hunk 3 above)
Comment (justschen) [reply src/vs/editor/contrib/codeAction/browser/codeActionContributions.ts:54]: "value of this configuration will be read only from trusted sources." not entirely sure what this means? our other code actions on save aren't `restricted` iirc, but unsure if this needs or not

File: src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts (hunk 2 above)
Comment (justschen) [reply src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts:289]: we don't atm

File: src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts (hunk 1 above)
Comment (justschen) [reply src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts]: yep left it in from when i wasn't using the helper func 👍

**Review by Tyriar — APPROVED** [review]

**Comment by Tyriar** [comment]
Thanks for the updates 👍
